File: algorithm_library/python/math/eratosthenes.py
# エラトステネスの篩
# O (N log log N)
def eratosthenes(N: int): 
  prime_number = list() # 素数
  isprime = [True] * (N+1) # isprime[i]==True なら i は素数
  isprime[0], isprime[1] = False, False
  for p in range(2, N+1):
      if not isprime[p]: continue
      q = p * 2
      while q <= N:
          isprime[q] = False
          q += p
  for p in range(2,N+1):
      if isprime[p]: prime_number.append(p)
  return prime_number, isprime

# M 以下の素数リストを返す
# O(M log log M)
def res_primeList(M):
    isprime = [True] * (M+1)
    isprime[0], isprime[1] = False, False
    for i in range(2, M+1):
        if i*i > M: break
        if isprime[i]:
            for j in range(i*i, M+1, i):
                isprime[j] = False
    return [i for i in range(2, M+1) if isprime[i]]

File: algorithm_library/cpp-from-py/math/eratosthenes.cpp
#include <bits/stdc++.h>
using namespace std;

pair<vector<int>, vector<bool>> eratosthenes(int N) {
    vector<int> prime_number;
    vector<bool> isprime(N + 1, true);
    if (N >= 0) isprime[0] = false;
    if (N >= 1) isprime[1] = false;
    for (int p = 2; p <= N; p++) {
        if (!isprime[p]) continue;
        for (long long q = 2LL * p; q <= N; q += p) isprime[q] = false;
    }
    for (int p = 2; p <= N; p++) if (isprime[p]) prime_number.push_back(p);
    return {prime_number, isprime};
}

vector<int> res_primeList(int M) {
    vector<bool> isprime(M + 1, true);
    if (M >= 0) isprime[0] = false;
    if (M >= 1) isprime[1] = false;
    for (int i = 2; i <= M && 1LL * i * i <= M; i++) {
        if (isprime[i]) for (long long j = 1LL * i * i; j <= M; j += i) isprime[j] = false;
    }
    vector<int> res;
    for (int i = 2; i <= M; i++) if (isprime[i]) res.push_back(i);
    return res;
}
